onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,288 @@
+  }
+  for (Node* consumer : consumer_nodes) {
+    if (consumer) {
+      std::vector<NodeArg*> consumer_inputs = consumer->MutableInputDefs();
+      auto consumer_outputs = consumer->MutableOutputDefs();
+      for (std::pair<NodeArg*, NodeArg*> p : substitute_list) {
+        std::replace(consumer_inputs.begin(), consumer_inputs.end(), p.first, p.second);
+      }
+      Node& new_consumer =  graph.AddNode(graph.GenerateNodeName(consumer->Name() + "_replace"),
+                                          consumer->OpType(),
+                                          "Created to replace a node",
+                                          consumer_inputs,
+                                          consumer_outputs,
+                                          &consumer->GetAttributes(),
+                                          consumer->Domain());
+      (void) new_consumer;
+      removed_nodes.push_front(consumer->Index());
+    }
+  }
+  return Status::OK();
+}
+
+static Status RemoveCastNode(Graph& graph, Node* cast, std::deque<onnxruntime::NodeIndex>& removed_nodes)
+{
+  NodeArg* cast_input = cast->MutableInputDefs()[0];
+  auto parent = graph.GetMutableProducerNode(cast_input->Name());
+  if (parent) {
+    graph_utils::FinalizeNodeFusion(graph, *parent, *cast);
+  } else {
+    removed_nodes.push_front(cast->Index());
+  }
+  return Status::OK();
+}
+
+// SearchUpstream:
+// Recursively traverse the graph upstream collecting all the NodeArgs that require a cast
+// inorder to remove an FP16 Cast operation down the graph.
+static void SearchUpstream(Graph& graph, NodeArg* node_arg, std::set<NodeArg*>& require_cast)
+{
+  Node* node = graph.GetMutableProducerNode(node_arg->Name());
+  if (node == nullptr) {
+    // The graph inputs don't have the producer nodes
+    require_cast.insert(node_arg);
+  } else {
+    std::string op_type = node->OpType();
+    if (std::find(fp16_allow.begin(), fp16_allow.end(), op_type) == fp16_allow.end() &&
+        std::find(fp16_safe.begin(), fp16_safe.end(), op_type) == fp16_safe.end()) {
+      require_cast.insert(node_arg);
+    } else {
+      for (NodeArg* node_input : node->MutableInputDefs()) {
+        SearchUpstream(graph, node_input, require_cast);
+      }
+    }
+  }
+}
+
+// SearchDownstream:
+// Recursively traverse the graph downstream collecting all the NodeArgs that require a cast
+// inorder to remove an FP32 Cast operation up the graph.
+static void SearchDownstream(Graph& graph, NodeArg* node_arg, std::set<NodeArg*>& require_cast)
+{
+  for (Node* node : graph.GetMutableConsumerNodes(node_arg->Name())) {
+    if (node) {
+      std::string op_type = node->OpType();
+      if (std::find(fp16_allow.begin(), fp16_allow.end(), op_type) == fp16_allow.end()) {
+        require_cast.insert(node_arg);
+      } else {
+        for (NodeArg* node_output : node->MutableOutputDefs()) {
+          SearchDownstream(graph, node_output, require_cast);
```

Comment:
ah... nice recursion! Haven't seen the use of recursion for a while... I am not too sure about if there is any guideline of using recursion in ORT. 

Alternatively, you can use a to_visit list to keep track of the NodeArg* that need to expand latter.

Reply:
The depth of recursion is determined by how far the propagation can go. 
In the extremely unlikely case, If the graph is large, and as we add more ops to the allow-list, the propagation might go pretty far. :)